securesystemslib/hsm.py

```diff
@@ -0,0 +1,448 @@
+
+    try:
+      public_key_objects = self.sess.findObjects([(PyKCS11.CKA_CLASS,
+          PyKCS11.CKO_PUBLIC_KEY)])
+    except:
+      raise securesystemslib.exceptions
+    return public_key_objects
+
+
+
+
+
+  def get_public_key_value(self, public_key_handle):
+    """
+    <Purpose>
+      Get the public key value corresponding to the 'public_key_handle'
+
+    <Arguments>
+      public_key_handle:
+        element of the list returne by get_public_key_object().
+
+    <Returns>
+      'cryptography' public key object
+    """
+
+    public_key_value = self.sess.getAttributeValue(public_key_handle,
+        [PyKCS11.CKA_VALUE])[0]
+    public_key_value = bytes(public_key_value)
+
+    # Public key value exported from the HSM is der encoded
+    public_key = serialization.load_der_public_key(public_key_value,
+        default_backend())
+    return public_key
+
+
+
+
+
+  def get_X509_objects(self):
+    """
+    <Purpose>
+      Get object handle of the X509 certificates stored on the HSM.
+
+    <Returns>
+      List of all the available certificate handles.
+    """
+
+    x509_objects = self.sess.findObjects([(PyKCS11.CKA_CLASS,
+        PyKCS11.CKO_CERTIFICATE)])
+    return x509_objects
+
+
+
+
+
+  def get_X509_value(self, x509_handle):
+    """
+    <Purpose>
+      Get the certificate value corresponding to the 'x509_handle'.
+
+    <Arguments>
+      x509_handle:
+        element from the list returned by get_X509_objects().
+
+    <Returns>
+      'cryptography' public key object.
+    """
+
+    x509_value = self.sess.getAttributeValue(x509_handle,
+        [PyKCS11.CKA_VALUE])[0]
+    x509_certificate = x509.load_der_x509_certificate(bytes(x509_value))
+
+    return x509_certificate
+
+
+
+
+
+  def generate_signature(self, data, private_key_handle):
+    """
+    <Purpose>
+      Calculate signature over 'data' using the private key corresponding
+      to the 'private_key_handle'
+
+      Supported Keys
+      1. RSA
+      2. ECDSA
+
+    <Arguments>
+      data:
+        bytes over which the signature is to be calculated
+        'data' should be encoded/serialized before it is passed here.
+
+      private_key_handle:
+        element form the list returned by the get_private_key_objects()
+
+    <Exceptions>
+      securesystemslib.exceptions.UnsupportedAlgorithmError, when the
+      key type of the 'private_key_handle' is not supported
+
+    <Returns>
+      HEX string of the generated signature
+    """
+
+    mechanism = None
+    key_type = self.sess.getAttributeValue(private_key_handle,
+        [PyKCS11.CKA_KEY_TYPE])[0]
+
+    if PyKCS11.CKK[key_type] == 'CKK_RSA':
+
+      # RSA-PSS with SHA256 hash to be used for signature generation
+      RSA_PSS = PyKCS11.CKM_SHA256_RSA_PKCS_PSS
+      # SHA256 hash to be used to digest the data
+      HASH = PyKCS11.CKM_SHA256
+      # Mask generating function
+      MGF = PyKCS11.CKG_MGF1_SHA256
+      # length of salt to be used for hashing
+      salt_length = 32
```

Comment:
You define this same value a few times. I would move the variable to the top of the file and make it all caps.

Reply:
If we plan to add other RSA signing mechanism in future, then it might be a problem. If not, I can move the variables to the top.

Reply:
I made the required changes!